### PR TITLE
Use blob store cache for Lucene compound files

### PR DESCRIPTION
**Pull request by tlrx**
The blob store cache is used to cache a variable length of the begining of Lucene files in the `.snapshot-blob-cache` system index. This is useful to speed up Lucene directory opening during shard recovery and to limit the number of bytes downloaded from the blob store when a searchable snapshot shard must be rebuilt. 

This pull request adds support for compound files segment (.cfs) when they are partially cached (ie, `Storage.SHARED_CACHE`) so that the files they are composed of can also be cached in the blob store cache index.

 Co-authored-by: Yannick Welsch yannick@welsch.lu

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java [hunk 1]
```diff
@@ -371,61 +339,6 @@ private void refreshSystemIndex() {
...
-                        file.getParent().getFileName().toString()
-                    ),
-                    INDEX_SHARD_SNAPSHOT_FORMAT.deserialize(fileName, xContentRegistry(), Streams.readFully(Files.newInputStream(file)))
-                );
-            }
-        }));
-        return Map.copyOf(blobsPerShard);
-    }
-
-    private void assertCachedBlobsInSystemIndex(final String repositoryName, final Map<String, BlobStoreIndexShardSnapshot> blobsInSnapshot)
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:396]: In a follow up I'll add tests to verify the exact cached documents for CFS and non-CFS Lucene files, but for this integration test I think it is sufficient to verify that no bytes were downloaded after the second mount.

File: test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java [hunk 2]
```diff
@@ -121,7 +121,7 @@ protected void doRun() throws Exception {
                                     clone = indexInput.clone();
                                 } else {
                                     final int sliceEnd = between(readEnd, length);
-                                    clone = indexInput.slice("concurrent slice (0, " + sliceEnd + ") of " + indexInput, 0L, sliceEnd);
+                                    clone = indexInput.slice("concurrent slice" + randomFileExtension(), 0L, sliceEnd);
```
Comment (tlrx) [inline test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java:124]: this is to better reflect what Lucene is doing when slicing a CFS file

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java [hunk 3]
```diff
@@ -134,32 +147,37 @@ private ByteRange computeRange(long position) {
...
-        final long position = getFilePointer() + this.offset;
+        final long position = getAbsolutePosition();
         final int length = b.remaining();
 
         logger.trace("readInternal: read [{}-{}] ([{}] bytes) from [{}]", position, position + length, length, this);
         try {
             final CacheFile cacheFile = cacheFileReference.get();
 
             // Can we serve the read directly from disk? If so, do so and don't worry about anything else.
+            if (isReadFromCompoundFileDuringRecovery(position, length) == false) {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:158]: This is to "force" the creation of cached blob docs during directory opening.

**Comment by tlrx** [comment]
Build failure in elasticsearch-ci/1 is tracked in #69980. I'm waiting for this test to be muted and I'll rerun the tests.

**Comment by martijnvg** [comment]
@tlrx I just muted it.

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java [hunk 4]
```diff
@@ -154,17 +148,14 @@ public void testBlobStoreCache() throws Exception {
...
-
         expectThrows(
             IndexNotFoundException.class,
             ".snapshot-blob-cache system index should not be created yet",
             () -> systemClient().admin().indices().prepareGetIndex().addIndices(SNAPSHOT_BLOB_CACHE_INDEX).get()
         );
 
-        Storage storage = randomFrom(Storage.values());
+        // TODO randomize this with FULL_COPY too when cold tier also handle blob cache for footers
+        final Storage storage = Storage.SHARED_CACHE;
```
Comment (ywelsch) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java:158]: I prefer that we keep the cold storage variant here as well, and continue to assert about cfs further down below.

**Review by ywelsch — COMMENTED** [review]
I've left one comment, o.w. looking good.

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java (hunk 4 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/blobstore/cache/SearchableSnapshotsBlobStoreCacheIntegTests.java]: Makes sense, I pusched 77828c94f6d8cf68c5b16f50dec94ff3c800aec9

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks for your help on this Yannick

**Pull request merged**